File: sdk/php/src/utils/ArrayHelpers.php
<?php

namespace SymbolSdk\Utils;

use SplFixedArray;
use ArrayAccess;
use OutOfRangeException;
use RangeException;

class ArrayHelpers {
	public static function deepCompare($lhs, $rhs)
	{
		if (!is_array($lhs) && !($lhs instanceof SplFixedArray) && !($lhs instanceof ArrayAccess)) {
			if ($lhs === $rhs)
				return 0;
	
			return $lhs > $rhs ? 1 : -1;
		}
	
		if (count($lhs) !== count($rhs))
			return count($lhs) > count($rhs) ? 1 : -1;
	
		for ($i = 0; $i < count($lhs); ++$i) {
			$compareResult = self::deepCompare($lhs[$i], $rhs[$i]);
			if ($compareResult !== 0)
				return $compareResult;
		}
	
		return 0;
	}
	
	public static function readArrayImpl(BinaryReader &$reader, callable $factoryMethod, $accessor, $shouldContinue)
	{
		$elements = [];
		$previousElement = null;
		$i = 0;
		while ($shouldContinue($i, $reader)) {
			$element = call_user_func($factoryMethod, $reader);
	
			if ($element->size() <= 0) {
				throw new RangeException('element size has invalid size');
			}
	
			if ($accessor && $previousElement && self::deepCompare($accessor($previousElement), $accessor($element)) >= 0) {
				throw new RangeException('elements in array are not sorted');
			}
	
			array_push($elements, $element);
			$previousElement = $element;
			++$i;
		}
	
		return $elements;
	}
	
	public static function writeArrayImpl(BinaryWriter $output, array $elements, int $count, $accessor)
	{
		for ($i = 0; $i < $count; ++$i) {
			$element = $elements[$i];
			if ($accessor && 0 < $i && 0 <= self::deepCompare($accessor($elements[$i - 1]), $accessor($element)))
				throw new OutOfRangeException('array passed to write array is not sorted');
	
			$output->write($element->serialize());
		}
	}
	
	public static function alignUp($size, $alignment)
	{
		return intval(($size + $alignment - 1) / $alignment) * $alignment;
	}
	
	public static function size($elements, $alignment = 0, $skipLastElementPadding = false): int
	{
		if ($elements == null) return 0;
		if ($alignment == 0) {
			return array_sum(array_map(function ($e) {
				return $e->size();
			}, $elements));
		}
	
		if (!$skipLastElementPadding) {
			return array_sum(array_map(function ($e) use ($alignment) {
				return self::alignUp($e->size(), $alignment);
			}, $elements));
		}
	
		$lastIndex = count($elements) - 1;
		$lastElement = $elements[$lastIndex];
		$sum = array_sum(array_map(function ($e) use ($alignment) {
			return self::alignUp($e->size(), $alignment);
		}, array_slice($elements, 0, $lastIndex)));
	
		$sum += $lastElement->size();
		return $sum;
	}
	
	public static function readArray($reader, callable $factoryMethod, $accessor = null)
	{
		return self::readArrayImpl($reader, $factoryMethod, $accessor, fn ($_, BinaryReader $reader) => 0 < $reader->getRemainingLength());
	}
	
	public static function readArrayCount($bufferInput, callable $factoryMethod, $count, $accessor = null)
	{
		return self::readArrayImpl($bufferInput, $factoryMethod, $accessor, function ($index) use ($count) {
			return $count > $index;
		});
	}
	
	public static function readVariableSizeElements(BinaryReader &$reader, callable $deserializeFunc, $binarySize, $alignment, $skipLastElementPadding = false)
	{
		$elements = [];
		while (0 < $binarySize) {
			$buffer = $reader->readRemaining();
			$element = call_user_func($deserializeFunc, $buffer);
	
			if (0 >= $element->size())
				throw new OutOfRangeException('element size has invalid size');
	
			array_push($elements, $element);
	
			$alignedSize = ($skipLastElementPadding && $element->size() >= $reader->getRemainingLength())
				? $element->size()
				: self::alignUp($element->size(), $alignment);
			if ($alignedSize > strlen($buffer))
				throw new OutOfRangeException('unexpected buffer length');
	
			$reader->advance($alignedSize);
			$binarySize -= $alignedSize;
		}
	
		return $elements;
	}
	
	public static function writeArray(BinaryWriter $output, array $elements, $accessor = null)
	{
		if ($elements == Null) return;
		self::writeArrayImpl($output, $elements, count($elements), $accessor);
	}
	
	public static function writeArrayCount(BinaryWriter $output, array $elements, int $count, $accessor = null)
	{
		if ($elements == Null) return;
		self::writeArrayImpl($output, $elements, $count, $accessor);
	}
	
	public static function writeVariableSizeElements(BinaryWriter $output, array $elements, int $alignment, $skipLastElementPadding = false)
	{
		foreach ($elements as $index => $element) {
			$output->write($element->serialize());
	
			if (!$skipLastElementPadding || count($elements) - 1 !== $index) {
				$alignedSize = self::alignUp($element->size(), $alignment);
	
				if ($alignedSize - $element->size() > 0) {
					$padding = str_repeat("\x00", $alignedSize - $element->size());
					$output->write($padding);
				}
			}
		}
	}	
}

File: sdk/csharp/SymbolSdk/Utils/ArrayHelpers.cs
namespace SymbolSdk;
public static class ArrayHelpers
{
    public static int DeepCompare(object lhs, object rhs)
    {
        if (lhs is not (Array or byte[]))
        {
            if (Equals(lhs, rhs))
                return 0;
            var property = lhs.GetType().GetProperty("Value");
            if (property == null)
            {
                return Comparer<object>.Default.Compare(lhs, rhs);
            }
            {
                var lhsValue = (uint)property.GetValue(lhs);
                var rhsValue = (uint)property.GetValue(rhs);
                return lhsValue > rhsValue ? 1 : -1;
            }
        }

        var lhsArray = (Array)lhs;
        var rhsArray = (Array)rhs;
        
        if (lhsArray.Length != rhsArray.Length)
            return lhsArray.Length > rhsArray.Length ? 1 : -1;

        return lhsArray.Cast<object>().Select((_, i) => 
            DeepCompare(lhsArray.GetValue(i), rhsArray.GetValue(i))).FirstOrDefault(compareResult => compareResult != 0);
    }
    
    private static void AddPadding(int size, BinaryWriter bw,
        int alignment)
    {
        var padding = GetPadding(size, alignment);
        while (padding > 0)
        {
            bw.Write((byte)0);
            padding--;
        }
    }

    private static int GetPadding(int size, int alignment)
    {
        if (alignment == 0)
        {
            return 0;
        }
        return 0 == size % alignment ? 0 : alignment - (size % alignment);
    }
    
    public static void WriteArray<T>(BinaryWriter bw, IEnumerable<T> elements) where T : ISerializer
    {
        foreach (var t in elements)
        {
            bw.Write(t.Serialize());
            AddPadding((int)t.Size, bw, 0);
        }
    }

    /**
	 * Calculates aligned size.
	 * @param {uint} size Size.
	 * @param {uint} alignment Alignment.
	 * @returns {uint} Size rounded up to alignment.
	 */
    private static uint AlignUp(uint size, uint alignment)
    {
        return (uint)Math.Floor((size + alignment - 1) / (float)alignment) * alignment;
    }
    
    public static uint Size<T>(T[] elements, uint alignment = 0, bool skipLastElementPadding = false) where T : ISerializer
    {
        if (!elements.Any()) return 0;
        if (alignment == 0) return (uint)elements.Sum(e => e.Size);

        if (!skipLastElementPadding) return (uint)elements.Sum(e => AlignUp(e.Size, alignment));

        return (uint)elements.Take(elements.Length - 1).Sum(e => AlignUp(e.Size, alignment)) +
               (uint)elements.Skip(elements.Length - 1).Sum(e => e.Size);
    }

    /**
	 * Writes array of variable size objects.
	 * @param {BinaryWriter} output An output sink.
	 * @param {array&lt;T&gt;} elements Serializable elements.
	 * @param {uint} alignment Alignment used to make sure each object is at boundary.
	 * @param {bool} skipLastElementPadding true if last element should not be aligned/padded.
	 */
    public static void WriteVariableSizeElements<T>(BinaryWriter bw, T[] elements, uint alignment, bool skipLastElementPadding = false) where T : ISerializer
    {
        for (var i = 0; i < elements.Length; i++)
        {
            bw.Write(elements[i].Serialize().ToArray());
            if (skipLastElementPadding && elements.Length - 1 == i) continue;
            var alignedSize = AlignUp(elements[i].Size, alignment);
            if (alignedSize != elements[i].Size)
            {
                bw.Write(new byte[alignedSize - elements[i].Size]);
            }
        }
    }

    /**
	 * Reads array of variable size objects.
	 * @param {BinaryReader} bufferInput A uint8 array.
	 * @param {Func} FactoryClass Factory used to deserialize objects.
	 * @param {uint} alignment Alignment used to make sure each object is at boundary.
	 * @param {bool} skipLastElementPadding true if last element is not aligned/padded.
	 * @returns {array&lt;T&gt;} Array of deserialized objects.
	 */
    public static T[] ReadVariableSizeElements<T>(BinaryReader br, Func<BinaryReader, T> factoryClass, uint payloadSize, uint alignment, bool skipLastElementPadding = false) where T : IStruct
    {
        var elements = new List<T>();
        if (alignment == 0) return elements.ToArray();
        var remainingByteSizes = (int)payloadSize;
        while (remainingByteSizes > 0)
        {
            var entity = factoryClass(br);
            elements.Add(entity);
            var alignedSize = (skipLastElementPadding && entity.Size >= br.BaseStream.Length - br.BaseStream.Position)
                ? (int)entity.Size
                : (int)AlignUp(entity.Size, alignment);
            if (alignedSize > remainingByteSizes)
                throw new Exception("unexpected buffer length");
            br.ReadBytes((int)(alignedSize - entity.Size));
            remainingByteSizes -= alignedSize;
        }
        return elements.ToArray();
    }

    /**
	 * Reads array of objects.
	 * @param {BinaryReader} bufferInput Input buffer.
	 * @param {Func} accessor Optional accessor used to check objects order.
	 * @returns {array&lt;T&gt;} Array of deserialized objects.
	 */
    public static T[] ReadArray<T>(BinaryReader br, Func<BinaryReader, T> factory) where T : ISerializer
    {
        var position = br.BaseStream.Position;
        var length = (int)br.BaseStream.Length;
        var elements = new List<T>();
        if (length == position) return elements.ToArray();
        while (br.BaseStream.Position < length)
        {
            var entity = factory(br);
            elements.Add(entity);
        }
        return elements.ToArray();
    }

    /**
	 * Reads array of deterministic number of objects.
	 * @param {BinaryReader} bufferInput A uint8 array.
	 * @param {Func} accessor Optional accessor used to check objects order.
     * @param {byte} count Number of object to deserialize.
	 * @returns {array&lt;T&gt;} Array of deserialized objects.
	 */
    public static T[] ReadArrayCount<T>(BinaryReader br, Func<BinaryReader, T> factory, byte count) where T : ISerializer
    {
        var alignment = br.BaseStream.Length;
        var elements = new List<T>();
        if (alignment == 0) return elements.ToArray();
        for (var i = 0; i < count; i++)
        {
            var entity = factory(br);
            elements.Add(entity);
        }
        return elements.ToArray();
    }
    
    public static object GetValue(object o) {
        var comparerMethod = o.GetType().GetMethod("Comparer"); 
        return comparerMethod != null ? comparerMethod.Invoke(o, new object[] { }) : o.GetType().GetField("Value").GetValue(o);
    }
}